### DISCUSSION TITLE
[tinytex] Frequently Asked Questions | /tinytex/faq/

### DISCUSSION THREAD

**Discussion by yihui** [rbind/yihui#1358] [Comments]
https://yihui.org/tinytex/faq/

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
If you have too many packages to install, I guess the best way is to install them on another computer that's not behind the firewall, and copy TinyTeX to your working laptop: <https://yihui.org/en/2018/08/tinytex-flash-drive/>

If you only need to install one or two packages, you can download them from CTAN, and unzip the packages to the TinyTeX installation dir. The most common dir to put these packages is file.path(tinytex::tinytex_root(), 'texmf-dist', 'tex', 'latex').

> Originally posted on 2020-02-11 15:13:42

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
First, run this command in your terminal (or command window) to get the names of installed LaTeX packages:

     tlmgr info --list --only-installed --data name

Then reinstall TinyTeX, and use

    tlmgr install 

followed by the package names (separated by spaces) to install the packages.

> Originally posted on 2020-04-15 21:41:50

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I'm investigating this issue right now. Please keep an eye on <https://github.com/yihui/tinytex/issues/210> Thanks, and I'm sorry for the trouble!

> Originally posted on 2020-05-14 03:03:09

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I don't what the problem could be, but if tinytex::install_tinytex() doesn't work, you may try tinytex:::install_prebuilt().

> Originally posted on 2020-07-04 18:18:57

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
It will be great if you could post the full original error message. Anyway, if you have difficulty in downloading [TinyTeX.zip](https://yihui.org/tinytex/TinyTeX.zip), you may download it using your web browser, and pass the path of the zip file to tinytex:::install_prebuilt() to install it.

> Originally posted on 2021-08-19 02:44:22

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
What happens when you tinytex::install_tinytex()?

> Originally posted on 2021-10-20 00:10:46

**Reply by yihui** [reply]
You can try tinytex::install_tinytex(version = 'latest') instead.

> Originally posted on 2021-10-20 13:20:11

**Reply by yihui** [reply]
No, this line of code only needs to run once. After you install TinyTeX, you are done. You don't need to reinstall it over and over again.

> Originally posted on 2021-10-20 13:42:17

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Sorry, but I can't do anything about it. This file is from TeX Live, and you may contact them via the mailing list: <https://www.tug.org/texlive/contribute.html>

> Originally posted on 2021-11-09 22:41:31

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Please install the latest version of RStudio: <https://github.com/rstudio/rmarkdown/issues/1990>

> Originally posted on 2021-12-22 15:58:29

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Unfortunately I don't think there is a robust solution to this problem. The complication comes from the large number of CTAN mirrors. Every year in early April when a new version of TeX Live is released, this problem will arise because it takes time (one or a few more days) for mirrors to be synced up. The mirror that you are currently using may or may not have gotten the latest version of TeX Live yet.

Back to your question, tinytex::tlmgr_update() should automatically detect if you have a too old version of TeX Live and emit a message if you do: <https://github.com/yihui/tinytex/blob/c3a71c7/R/tlmgr.R#L223-L231> The message also tells you that you may have to wait for some time before you can tinytex::reinstall_tinytex().

> Originally posted on 2022-04-14 03:54:02

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
You can report this issue to <https://github.com/quarto-dev/quarto-cli/discussions> or <https://github.com/rstudio/rstudio/issues>. I'm not the right person to help with the RStudio IDE issue, since I don't develop the IDE. Thanks!

> Originally posted on 2022-09-12 04:21:25

[a comment by an automated account is not shown]

**Comment by littleshark-xuan** [comment]
When I want to knit pdf file in Rstudio,it always error that:! LaTeX Error: Unicode character 长 (U+957F)
               not set up for use with LaTeX.
But all my files are in English, I tried many ways to still have this problem, even after downloading someone else's new rmd file, it's still the same, I also uninstalled and reinstalled R many times. Even reset the computer still has this problem

**Comment by MWS-Koewi** [comment]
Hello,
I need to set up an R infrastructure on Windows clients in an internal company network where there is no connection to the Internet. Basically this is not a problem. I have pulled all the R packages to an internal ftp server and changed the config of R so that all packages can only be pulled from this server.
The whole thing should now be extended with the TinyTex tool. Unfortunately I haven't found any info yet how to implement this.
Maybe I have to say that I can't run "normal" installation packages, but everything that is needed is installed on a "naked" machine and the IT department then determines what everything was installed where and then builds a separate MSI package from it, which is made available in the company's internal software center. 

The question that arises now is, how can I best map this? As I said, I have the possibility to store reloadable packages on an ftp server.

**Reply by yihui** [reply]
For the company environment, I'd recommend that you install the full TinyTeX unless you lack the 4Gb disk space: https://yihui.org/en/2022/05/tinytex-full/

**Reply by MWS-Koewi** [reply]
With the version, the TinyTex-1 version would probably be enough for me. The question was rather how exactly do I install this on the reference machine so that all packages are then captured and included in the later msi installation?

**Reply by yihui** [reply]
I don't know how the MSI package is made, but TinyTeX is entirely portable. No matter where you put it on the reference machine, you can use it. You may see https://yihui.org/en/2018/08/tinytex-flash-drive/ for more info.

**Comment by JJGabe** [comment]
Greetings, Yihui. I am encountering several issues but I think they are part of the same problem. I am running Ubuntu 22.04 and I originally installed texlive-full through the command line for use with TexStudio as I do all of my article writing there. However, I'd like to use RMarkdown for managing my CV so I installed tinytex using install.packages('tinytex'). The main issue I am dealing with now, which I believe started with the update to R 4.3.2, is that my awesome-cv is causing an error when I try to include a bibtex file. All the rest of my CV entries (education, certifications, experience) compile with no issues but when I add a block with `bibliography_entries("input-docs/My-Pubs.bib")` I get a rather generic error message saying `! LaTeX Error: Something's wrong--perhaps a missing \item`. In the awesome-cv.cls I cannot find the specific code that deals with the bibliography and there is only one section (\newenvironment{cvitems}, line 697) that uses `begin{itemize}` where I would expect `\item` to be used.

Going through your debugging steps, I think the issue is where RStudio is looking for tinytex. When I run `tinytex::tinytex_root()` I get an error message `Error in root_dir(normalizePath(path), "..", "..", "..") : /usr/share/texlive does not seem to be the root directory of TeX Live (no "bin/" dir under it)`. Prior to the R update when my CV was compiling perfectly, I had tinytex in `~/R/x86_64-pc-linux-gnu-library/4.3` along with `~/.TinyTeX` (this second file is empty). This error also comes up when I use `tinytex::uninstall_tinytex()` so I am not able to reinstall it. Do you have any solutions to this problem?

Thanks,
Jeremy

**Reply by yihui** [reply]
Hi Jeremy, I don't know the cause of the LaTeX error, but it seems you don't really have TinyTeX installed. You said `~/.TinyTeX` was empty, so `uninstall_tinytex()` won't work (there is nothing to uninstall). If you want to use TinyTeX, you need to install it first: `tinytex::install_tinytex()`.

**Comment by JJGabe** [comment]
Thanks, but that hasn't quite solved my problem. I ran `tinytex::install_tinytex(dir = "~/TinyTex")` (which was not an existing folder) and before running it gave me a message that there was already a tlmgr folder found because I had installed texlive through the terminal before using it in RStudio. However, RStudio is still not recognising the correct path to TinyTex since when I run `tinytex::uninstall_tinytex()` I get the same error as before, that the root_dir is wrong. How can I set the tinytex path to `~/TinyTex` or `~/bin`, which was created with the tinytex installation and contains the tlmgr link?

**Reply by JJGabe** [reply]
I also get a similar error when trying to run `tinytex::tlmgr_update()`. It says that `directory /var/lib/texmf/web2c is not writable` (I assume this is where my TeXStudio is linked) even after I ran `tinytex::tlmgr_install()`

**Reply by yihui** [reply]
On Linux, TinyTeX should make symlinks from its installation directory (no matter where it is) to `~/bin`. As long as `~/bin` is in your `PATH` variable, both the **tinytex** R package and RStudio should find it. You can check `Sys.getenv('PATH')` in R or `echo $PATH` in your terminal.

If `~/bin` is not in your `PATH` (normally it should be), you could symlink TinyTeX's executables to a directory in `PATH`, such as `/usr/local/bin` (this one is almost guaranteed to be present in `PATH`), but I don't quite recommend this approach unless you `apt remove texlive-full` from your system, otherwise you may have conflicts between TinyTeX and TeX Live.

```sh
cd ~/TinyTeX/bin/*  # assuming you installed TinyTeX to ~/TinyTeX; the default is ~/.TinyTeX
./tlmgr path remove  # in case symlinks have been made
./tlmgr option sys_bin /usr/local/bin
sudo ./tlmgr path add
```

**Reply by JJGabe** [reply]
Well then I'm not sure what the issue is. As you can see from the output below, `~/bin` is the second directory in `Sys.getenv('PATH')` but when I run `tinytex::uninstall_tinytex()` right after, I get the root_dir error again.

```
> Sys.getenv('PATH')
[1] "/usr/bin:/home/jeremy/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/usr/lib/rstudio/resources/app/bin/quarto/bin:/usr/lib/rstudio/resources/app/bin/postback"
> tinytex::uninstall_tinytex()
Error in root_dir(normalizePath(path), "..", "..", "..") : 
  /usr/share/texlive does not seem to be the root directory of TeX Live (no "bin/" dir under it)
```

Not sure if you have any suggestions other than removing both texlive-full and any tinytex-related folders and then reinstalling. If I do go that route, do you have a suggested order of installing? Should I do tinytex first and then try to link my TeXStudio?

Thanks for your help

**Reply by JJGabe** [reply]
Sorry, that code block didn't come out properly

**Reply by yihui** [reply]
Do you actually have any symlinks under `~/bin`? If you installed TinyTeX successfully, there should be several symlinks under `~/bin`.

You can install `texlive-full` and TinyTeX simultaneously, although it shouldn't be necessary. `~/bin` has quite high precedence in `PATH`, so TinyTeX should be chosen when `texlive-full` is also installed.

**Reply by JJGabe** [reply]
Yep. It seems that folder is fully populated, there are a total of 58 files in there, including the links for latex, lualatex, pdflatex, and tlmgr. I tried to link my TexStudio to those symlinks (they were linked to symlinks in the `/usr/bin/` folder) but it failed to compile a document since it was not able to find all the packages installed with `texlive-full` (I originally installed this with `sudo apt install texlive-full`). Is there a way that I can force TinyTex to link to the `/usr/bin/` folder instead to see if that solves my problem?

The strange thing about the error message I'm getting is that the directory `/usr/share/texlive` that it is trying to find the `bin/` folder in doesn't even show up in the list of `Sys.getenv('PATH')`

**Reply by yihui** [reply]
> Is there a way that I can force TinyTex to link to the `/usr/bin/` folder instead to see if that solves my problem?

Yes, I have told you how to do it in a previous reply: https://github.com/yihui/yihui.org/discussions/1358#discussioncomment-7777546

**Reply by JJGabe** [reply]
Yes, you did provide a solution but you recommended against it unless I uninstalled texlive. However, if I uninstall texlive, then wouldn't the symlinks in the `/usr/bin/` folder be removed? So there wouldn't be any reason to try to link tinytex to that folder?

I appreciate the help and suggestions but there seems to be a major conflict on my machine between texlive and tinytex and since I do all of my writing in TexStudio, I will probably abandon the tinytex for now. I was really only interested in the programmatic CV aspect but I'm not even able to Knit the awesome-cv template right now.

**Reply by yihui** [reply]
After looking at your `PATH` variable agin, I think I understand the situation now:

```
/usr/bin:/home/jeremy/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:...
```

I don't know why `/usr/bin` appeared twice in your `PATH` (it's probably a problem of RStudio IDE). I don't have Ubuntu on my computer to test with, but I think `texlive-full`'s symlinks are made to `/usr/bin`, which has higher precedence over `~/bin`. Therefore **tinytex** cannot find TinyTeX (it detects `texlive-full` first).

For now, you can adjust your `PATH` via:

```r
Sys.setenv(PATH = gsub('^/usr/bin:', '', Sys.getenv('PATH')))
```

In the mean time, I'll see if I can improve the detection of TinyTeX in the **tinytex** package.

Another solution is to uninstall `texlive-full`. I'm reasonably sure TinyTeX alone can work well for you. You don't need both, and Ubuntu's `texlive` package is often too old and hard to update (which is one of the reasons why we created TinyTeX).

---

Update: Actually if you don't change the default installation path of TinyTeX (`~/.TinyTeX`), it should be correctly detected, too. No need to tweak `PATH` or uninstall `texlive`.

**Comment by memeplex** [comment]
I use TinyTeX now but have used BasicTeX in the past, installing it with brew and adding packages with tlmgr. Just out of curiosity, do the caveats you mention in https://yihui.org/tinytex/faq/#faq-15 also apply to BasicTeX brew setups?

**Reply by yihui** [reply]
The first caveat may not apply. I think BasicTex manages its own `PATH`: https://github.com/Homebrew/homebrew-cask/blob/30c5b09cf49e884848c65ce96dcdcda288c28fa1/Casks/b/basictex.rb#L30 It probably doesn't write symlinks to `/usr/local/bin`. I have never tested it so I'm not entirely sure.

The second caveat depends on where `tlmgr install` installs packages for BasicTeX., which I don't know. If packages are installed to a place that Homebrew doesn't clean up, they may survive the update.

**Reply by memeplex** [reply]
Thanks! I'm going to take a look to see what it actually does.

**Comment by dkStevensNZed** [comment]
I'm having trouble with tinytex/tlmgr under Windows 10. Working from RStudio, your code

```
tinytex::tlmgr_update()
tlmgr update --all --self
```
runs without comment but 
```
tinytex:::install_yihui_pkgs()
```
will list a bunch of LaTex packages but then quits with

```
tlmgr.pl: The TeX Live versions supported by the repository
https://ftp.tu-chemnitz.de/pub/tug/historic/systems/texlive/2021/tlnet-final
  (2016--2021)
do not include the version of the local installation
  (2023).
```
I'm not sure where to go next but the upshot is that I'm unable to use tlmgr to add any packages. My particular case now is trying to add a new presentation theme for beamer that is not in the 'included' list. Any advice would be welcome.

**Reply by yihui** [reply]
That problem might be transient. If you retry and it still doesn't work, you can try to set a different CTAN repository via `tinytex::tlmgr_repo()`.

**Comment by dkStevensNZed** [comment]
Thanks. This is an ongoing issue for me. Til now I've just worked around it. I'll try a different repository and report back.

**Comment by dkStevensNZed** [comment]
I switch to U of Utah's repository, and things seem to work now. Thanks for the suggestion!

**Comment by hossam-houssien** [comment]
could i use it with `pandoc`  as a pdf_engine

**Reply by yihui** [reply]
Yes. Anything that works with TeX Live should work with TinyTeX, too.
